test/integration2/migrate/migration-lifecycle-integration.spec.js

```diff
@@ -0,0 +1,303 @@
+            expect(afterEach.called).to.be.false;
+            expect(afterAll.called).to.be.false;
+          });
+        });
+
+        describe('afterAll', function () {
+          it('runs after the migrations batch', async function () {
+            let count;
+            await knex.migrate.latest({
+              directory: 'test/integration2/migrate/test',
+              afterAll: async (knexOrTrx) => {
+                const data = await knexOrTrx('knex_migrations').select('*');
+                count = data.length;
+              },
+            });
+
+            const data = await knex('knex_migrations').select('*');
+            expect(data.length).to.equal(count);
+          });
+
+          it('is not called if the migration fails', async function () {
+            const afterAll = sinon.stub();
+
+            await knex.migrate
+              .latest({
+                directory: 'test/integration2/migrate/test_with_invalid',
+                afterAll,
+              })
+              .catch(() => {});
+
+            expect(afterAll.called).to.be.false;
+          });
+        });
+
+        describe('beforeEach', function () {
+          it('runs before each migration', async function () {
+            const tableExistenceChecks = [];
+            const beforeEach = sinon.stub().callsFake(async (trx) => {
+              const hasFirstTestTable = await trx.schema.hasTable(
+                'migration_test_1'
+              );
+              const hasSecondTestTable = await trx.schema.hasTable(
+                'migration_test_2'
+              );
+              tableExistenceChecks.push({
+                hasFirstTestTable,
+                hasSecondTestTable,
+              });
+            });
+
+            await knex.migrate.latest({
+              directory: 'test/integration2/migrate/test',
+              beforeEach,
+            });
+
+            expect(beforeEach.callCount).to.equal(2);
+            expect(tableExistenceChecks).to.deep.equal([
+              {
+                hasFirstTestTable: false,
+                hasSecondTestTable: false,
+              },
+              {
+                hasFirstTestTable: true,
+                hasSecondTestTable: false,
+              },
+            ]);
+          });
+
+          it('does not run the migration or the afterEach/afterAll hooks if the hook fails', async function () {
+            const beforeEach = sinon
+              .stub()
+              .throws(new Error('force beforeEach hook failure'));
+            const afterEach = sinon.stub();
+            const afterAll = sinon.stub();
+
+            await knex.migrate
+              .latest({
+                directory: 'test/integration2/migrate/test',
+                beforeEach,
+                afterEach,
+                afterAll,
+              })
+              .catch((error) => {
+                expect(error.message).to.equal('force beforeEach hook failure');
```

Comment:
If the function won't reject this won't be called, saved the error in a value as well as if rejected or not